File: go/buildpy/internal/models/project.go
package models

import (
	"os"
	"path/filepath"

	"github.com/charmbracelet/log"
	"github.com/shakfu/buildpy/internal/shell"
)

type Project struct {
	Cwd       string
	Build     string
	Downloads string
	Src       string
	Install   string
}

func NewProject() *Project {
	cwd, err := os.Getwd()
	if err != nil {
		log.Fatal(err)
	}
	var build = filepath.Join(cwd, "build")
	var downloads = filepath.Join(build, "downloads")
	var src = filepath.Join(build, "src")
	var install = filepath.Join(build, "install")
	return &Project{cwd, build, downloads, src, install}
}

func (p *Project) Setup() {
	shell.Makedirs(p.Build, p.Downloads, p.Src, p.Install)
}

func (p *Project) Clean() {
	os.RemoveAll(p.Src)
}

func (p *Project) Reset() {
	os.RemoveAll(p.Build)
}

File: rs/buildpy/src/core/project.rs
use std::env;
use std::path::PathBuf;

use crate::ops::shell;
use crate::ops::log;

#[derive(Debug)]
pub struct Project {
    pub cwd: PathBuf,
    pub build: PathBuf,
    pub downloads: PathBuf,
    pub src: PathBuf,
    pub install: PathBuf,
}

impl Project {
    pub fn new() -> Self {
        let _cwd = env::current_dir().unwrap();
        let _build = _cwd.join("build");
        Self {
            cwd: _cwd,
            build: _build.clone(),
            downloads: _build.join("download"),
            src: _build.join("src"),
            install: _build.join("install"),
        }
    }

    pub fn setup(&self) {
        for p in &[
            self.build.as_path(),
            self.downloads.as_path(),
            self.src.as_path(),
            self.install.as_path(),
        ] {
            match std::fs::create_dir_all(p) {
                Ok(_) => log::info!("created: {}", p.display()),
                Err(e) => log::error!("failure: {e}"),
            };
        }
    }

    pub fn clean(&self) {
        shell::remove(self.src.as_path());
    }

    pub fn reset(&self) {
        shell::remove(self.build.as_path());
    }
}
